hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+        self.token.cfg().write(|w| w.eomdis().bit(disable_eom));
+    }
+
+    /// Write Back Disable
+    ///
+    /// * If set to false, Write Back Operations are permitted
+    /// * If set to true, Write Back Operations are forbidden, `CDWBN` bit is
+    ///   internally set to 1 and cannot be modified by a linked list element.
+    ///   The `CDWBN` bit of the `RCFG` structure member has no effect.
+    #[inline]
+    pub fn set_wbdis(&self, disable_eom: bool) {
+        self.token.cfg().write(|w| w.wbdis().bit(disable_eom));
+    }
+
+    // Security and tamper settings
+
+    /// Set Undefined Register Access Detection interrupt enable
+    #[inline]
+    pub fn set_urad_int(self) {
+        self.token.ier().write(|w| unsafe { w.rsu().bits(1) });
+    }
+
+    /// Disable Undefined Register Access Detection interrupt enable
+    #[inline]
+    pub fn disable_urad_int(self) {
+        self.token.idr().write(|w| unsafe { w.rsu().bits(1) });
+    }
+
+    /// Get Undefined Register Access Detection interrupt mask
+    #[inline]
+    pub fn get_urad_int_mask(&self) -> bool {
+        self.token.imr().read().urad().bits()
+    }
+
+    /// Get Undefined Register Access Detection interrupt status
+    #[inline]
+    pub fn get_urad_int(&self) -> bool {
+        self.token.isr().read().urad().bits()
+    }
+
+    /// Get Undefined Register Access Trace
+    ///
+    /// This field is only reset by `swrst`
+    #[inline]
+    pub fn get_urat(&self) -> Variant<u8, URAT_A> {
+        self.token.uasr().read().urat().variant()
+    }
+
+    /// When reading the interrupt (ISR) register, it is cleared
+    ///
+    /// This might result in unexpected results for example
+    /// when sequentially trying to determine which interrupt triggered.
+    ///
+    /// This is an alternative, return all the data from the register
+    /// and parse later with the designated `get_[name]_int` functions.
+    #[inline]
+    pub fn get_interrupt_status(&mut self) -> IcmInterrupt {
+        let interrupt_vector = self.token.isr().read().bits();
+        IcmInterrupt { interrupt_vector }
+    }
+    /// Trigger recalculation of memory monitor region specified
+    /// by the bitmask:
+    /// 0b0001 = region0
+    /// 0b0010 = region1
+    /// 0b0100 = region2
+    /// 0b1000 = region3
+    /// ...
+    /// 0b1111 = all regions
+    #[inline]
+    pub fn trigger_rehash(&self, bitmask: u8) {
+        self.token
+            .ctrl()
+            .write(|w| unsafe { w.rehash().bits(bitmask) });
+    }
+}
+
+/// Trait providing numerical identifier and
+/// offset for each ICM Region
+///
+/// ICM supports 4 memory regions
+pub trait IcmRegionNum: Sealed {
+    /// Numerical ID of the memory region
+    const NUM: usize;
+    /// Memory offset
+    const OFFSET: u32;
+}
+
+/// ICM Region 0
+pub enum Region0 {}
+impl Sealed for Region0 {}
+impl IcmRegionNum for Region0 {
+    const NUM: usize = 0;
+    const OFFSET: u32 = 0;
+}
```

Comment:
You could save some typing by turning this into a `seq!` macro call. We already use it throughout the HAL.

```rust
seq!(for N in 0..=3 {
    paste! {
        #[doc = "ICM Region " N]
        pub enum Region#N {}
        impl Sealed for Region#N {}
        impl IcmRegionNum for Region#N {
            const NUM: usize = N;
            const OFFSET: u32 = 0x10 * N;
        }
    }
});
```

Reply:
Very kind of you to provide a copy-paste solution, just added some `#[allow(clippy::...)]` to make Clippy happy.

Reply:
📎😃